src/libsystemd-network/ndisc-option.c

```diff
@@ -0,0 +1,850 @@
+                uint8_t prefixlen,
+                const struct in6_addr *address,
+                usec_t valid_lifetime,
+                usec_t preferred_lifetime) {
+
+        assert(options);
+        assert(prefixlen <= 128);
+        assert(address);
+
+        sd_ndisc_option *p = ndisc_option_new(SD_NDISC_OPTION_PREFIX_INFORMATION, offset);
+        if (!p)
+                return -ENOMEM;
+
+        p->prefix = (sd_ndisc_prefix) {
+                .flags = flags,
+                .prefixlen = prefixlen,
+                .address = *address,
+                .valid_lifetime = valid_lifetime,
+                .preferred_lifetime = preferred_lifetime,
+        };
+
+        in6_addr_mask(&p->prefix.address, p->prefix.prefixlen);
+
+        return ndisc_option_consume(options, p);
+}
+
+static int ndisc_option_parse_prefix(size_t offset, size_t len, const uint8_t *opt, Set **options) {
+        const struct nd_opt_prefix_info *pi = (const struct nd_opt_prefix_info*) ASSERT_PTR(opt);
+
+        assert(options);
+
+        if (len != sizeof(struct nd_opt_prefix_info))
+                return -EBADMSG;
+
+        if (pi->nd_opt_pi_type != SD_NDISC_OPTION_PREFIX_INFORMATION)
+                return -EBADMSG;
+
+        if (pi->nd_opt_pi_prefix_len > 128)
+                return -EBADMSG;
+
+        if (in6_addr_is_link_local(&pi->nd_opt_pi_prefix))
+                return -EBADMSG;
+
+        usec_t valid = be32_sec_to_usec(pi->nd_opt_pi_valid_time, /* max_as_infinity = */ true);
+        usec_t pref = be32_sec_to_usec(pi->nd_opt_pi_preferred_time, /* max_as_infinity = */ true);
+        if (pref > valid)
+                return -EBADMSG;
+
+        /* We only support 64 bits interface identifier for addrconf. */
+        uint8_t flags = pi->nd_opt_pi_flags_reserved;
+        if (FLAGS_SET(flags, ND_OPT_PI_FLAG_AUTO) && pi->nd_opt_pi_prefix_len != 64)
+                flags &= ~ND_OPT_PI_FLAG_AUTO;
+
+        return ndisc_option_add_prefix(options, offset, flags, pi->nd_opt_pi_prefix_len, &pi->nd_opt_pi_prefix, valid, pref);
+}
+
+int ndisc_option_add_redirected_header(Set **options, size_t offset, const struct ip6_hdr *hdr) {
+        assert(options);
+        assert(hdr);
+
+        sd_ndisc_option *p = ndisc_option_new(SD_NDISC_OPTION_REDIRECTED_HEADER, offset);
+        if (!p)
+                return -ENOMEM;
+
+        /* For safety, here we copy only IPv6 header. */
+        memcpy(&p->hdr, hdr, sizeof(struct ip6_hdr));
+
+        return ndisc_option_consume(options, p);
+}
+
+static int ndisc_option_parse_redirected_header(size_t offset, size_t len, const uint8_t *opt, Set **options) {
+        assert(opt);
+        assert(options);
+
+        if (len < sizeof(struct nd_opt_rd_hdr) + sizeof(struct ip6_hdr))
+                return -EBADMSG;
+
+        if (opt[0] != SD_NDISC_OPTION_REDIRECTED_HEADER)
+                return -EBADMSG;
+
+        sd_ndisc_option *p = ndisc_option_new(SD_NDISC_OPTION_REDIRECTED_HEADER, offset);
```

Comment:
Looks like there is a memory leak here (and `p` isn't used by this function as far as I can see):
```sh
=================================================================
==259==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 560000 byte(s) in 10000 object(s) allocated from:
    #0 0x7fae7f0eb1bc in __interceptor_calloc (/usr/lib64/clang/16/lib/libclang_rt.asan-x86_64.so+0xeb1bc) (BuildId: 2a39bb4aa6d13ff866a4a47997cefdbe31a0350f)
    #1 0xad5b10 in ndisc_option_new /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-option.c:57:30
    #2 0xadbe98 in ndisc_option_parse_redirected_header /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-option.c:322:30
    #3 0xada07d in ndisc_parse_options /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-option.c:796:29
    #4 0xa76fab in ndisc_router_parse /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc-router.c:119:13
    #5 0xa6c9a8 in ndisc_handle_router /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:212:13
    #6 0xa6c0cb in ndisc_recv /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:261:24
    #7 0x7fae7e73b2bc in source_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4187:21
    #8 0x7fae7e7392e4 in sd_event_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4808:21
    #9 0x7fae7e73eddd in sd_event_run /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4869:21
    #10 0x7fae7e73fd69 in sd_event_loop /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4891:21
    #11 0x416f6d in run /home/vagrant/systemd-asan/build/../src/network/networkd.c:109:13
    #12 0x413ffd in main /home/vagrant/systemd-asan/build/../src/network/networkd.c:116:1
    #13 0x7fae7cc49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #14 0x7fae7cc49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #15 0x412a14 in _start (/usr/lib/systemd/systemd-networkd+0x412a14) (BuildId: 619afa9520903783d507f9c515898f3f7d7315e2)
SUMMARY: AddressSanitizer: 560000 byte(s) leaked in 10000 allocation(s).
systemd-networkd.service: Child 259 belongs to systemd-networkd.service.
```